### PR TITLE
[FSDE-13889] Create e2e tests for bulk actions

### PR DESCRIPTION
**Pull request by josecarlosgonzalezv**
@evansito @soujanreddy

### CODE REVIEW

**Comment by SCMAUTO** [comment]
You can check this PRs instance in https://bf053ebce0db8a1a9a3186257d029524996f6d88-chi-pr.enterprise.ctl.io (external) and in https://nginx-pr-314-ux-chi.kubeodc-test.corp.intranet/ (internal)

File: cypress/integration/chi-vue/data-table.spec.js [hunk 1]
```diff
@@ -60,6 +60,10 @@ const DATA_TABLE_EVENTS = {
   EXPANSION: {
     EXPANDED: 'chiRowExpanded',
     COLLAPSED: 'chiRowCollapsed'
+  },
+  BULK_ACTIONS: {
```
Comment (evansito) [inline cypress/integration/chi-vue/data-table.spec.js:64]: I'd name this BULK_ACTIONS_EVENTS

File: cypress/integration/chi-vue/data-table.spec.js [hunk 2]
```diff
@@ -349,22 +365,21 @@ describe('Data Table', () => {
...
-          .first()
-          .as('trigger')
-          .trigger('mouseenter')
-          .then(() => {
-            cy.get('@trigger')
-              .find(`.${TOOLTIP_CLASSES.TOOLTIP}`)
-              .should(assertion);
-          });
-      });
+      it(`Tooltip element ${index} should ${!isVisible ? 'not' : ''
```
Comment (evansito) [inline cypress/integration/chi-vue/data-table.spec.js:368]: i'd name this in a more descriptive way, e.g. existsAssertion, or isVisibleAssertion

File: cypress/integration/chi-vue/data-table.spec.js [hunk 3]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
         });
     });
   });
+
+  describe.only('Bulk Actions', () => {
```
Comment (evansito) [inline cypress/integration/chi-vue/data-table.spec.js:1194]: .only should be removed

File: cypress/integration/chi-vue/data-table.spec.js [hunk 4]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+        .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.BUTTONS} 
+          .${BULK_ACTIONS_CLASSES.BUTTONS_MOBILE}`)
+        .should('not.be.visible');
+    });
+
+    it('Should go to page 2 and selected number of rows should be the same number', () => {
+      const selected = '1';
+
+      cy.get('@selectables')
+        .eq(0)
```
Comment (evansito) [inline cypress/integration/chi-vue/data-table.spec.js:1378]: i suggest using .first() instead for readability

File: cypress/integration/chi-vue/data-table.spec.js [hunk 5]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+        .click()
+        .then(() => {
+          cy.get(`[data-cy='data-table-bulk-actions']`)
+            .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.START} 
+                  .${BULK_ACTIONS_CLASSES.LABEL}`)
+            .contains(`Actions (${selected} Selected)`);
+        });
+      cy.get(`@pagination`)
+        .find(`.${ICON_BUTTON}`)
+        .eq(0)
```
Comment (evansito) [inline cypress/integration/chi-vue/data-table.spec.js:1395]: i suggest using .first() instead for readability

File: cypress/integration/chi-vue/data-table.spec.js [hunk 6]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+          cy.get('@selectables')
+            .eq(1)
+            .click();
+          cy.get(`[data-cy='data-table-bulk-actions']`)
+            .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.START} 
+                .${BULK_ACTIONS_CLASSES.LABEL}`)
+            .contains(`Actions (${selected} Selected)`);
+          cy.get(`@pagination`)
+            .find(`.${ICON_BUTTON}`)
+            .eq(0)
```
Comment (evansito) [inline cypress/integration/chi-vue/data-table.spec.js:1416]: i suggest using .first() instead for readability

File: test/chi/chi-vue/_data-table-base.pug [hunk 7]
```diff
@@ -137,8 +137,36 @@ h3 Description
...
+            chi-button(variant='flat' type='icon' aria-label='Compose')
+              chi-icon(icon='compose')
+            chi-button(variant='flat' type='icon' aria-label='Delete')
+              chi-icon(icon='delete')
+            chi-button(variant='flat' type='icon' aria-label='Print')
+              chi-icon(icon='print')
+          .chi-bulk-actions__buttons-desktop
+            chi-button(size='xs' aria-label='Download')
+              chi-icon(icon='arrow-to-bottom')
+              span Download
```
Comment (evansito) [inline test/chi/chi-vue/_data-table-base.pug:158]: we could use v-for loops here

File: test/chi/chi-vue/_data-table-base.pug [hunk 8]
```diff
@@ -1639,6 +1667,77 @@ script.
...
+          status: { label: 'Status', sortable: true, description: { title: 'Status', template: 'statusdesc', 
+            payload: { 
+              overdue: { status: "Overdue", icon: "circle-alert", color: "danger" }, 
+              due: { status: "Due", icon: 'warning', color: 'warning' }
+              } 
+            }
+          },
+          lastLogin: { label: "Last Login" },
+        },
+        body: [
```
Comment (evansito) [inline test/chi/chi-vue/_data-table-base.pug:1706]: the body data looks similar to exampleBody so you may use that

**Review by evansito — CHANGES_REQUESTED** [review]

File: cypress/integration/chi-vue/data-table.spec.js (hunk 1 above)
Comment (josecarlosgonzalezv) [reply cypress/integration/chi-vue/data-table.spec.js:64]: I am using the same naming as the other events, e.g. TOOLBAR does not have TOOLBAR_EVENTS:

image

File: cypress/integration/chi-vue/data-table.spec.js [hunk 9]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+                .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.START} 
+                  .${BULK_ACTIONS_CLASSES.LABEL}`)
+                .contains(`Actions (${selected} Selected)`);
+              cy.get("@rows").should("have.length", 3);
+              expect(spy).to.be.calledOnce;
+            });
+        });
+    });
+
+    it('Should close bulk actions and show data table is still in its original state', () => {
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:1291]:
```suggestion
    it('Should close bulk actions and show data table in its original state when clicking the X button', () => {
```

File: cypress/integration/chi-vue/data-table.spec.js [hunk 10]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+        .its('dataTableBulkActionsExample')
+        .then(dataTableBulkActionsExample => {
+          const spy = cy.spy();
+          const dataTableRef = dataTableBulkActionsExample.$refs.dataTableBulkActionsRef;
+          const paginationResults = '6';
+
+          dataTableRef.$on(`${DATA_TABLE_EVENTS.BULK_ACTIONS.CANCEL}`, spy);
+          cy.get(`[data-cy='data-table-bulk-actions']`)
+            .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.END}`)
+            .click()
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:1302]: Is this clicking the actual X button? Seems like the click happens on `.${BULK_ACTIONS_CLASSES.END}`.

File: cypress/integration/chi-vue/data-table.spec.js [hunk 11]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+        .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.BUTTONS} 
+          .${BULK_ACTIONS_CLASSES.BUTTONS_DESKTOP}`)
+        .should('be.visible');
+      cy.get(`[data-cy='data-table-bulk-actions']`)
+        .find(`.${BULK_ACTIONS_CLASSES.BULK_ACTIONS} .${BULK_ACTIONS_CLASSES.BUTTONS} 
+          .${BULK_ACTIONS_CLASSES.BUTTONS_MOBILE}`)
+        .should('not.be.visible');
+    });
+
+    it('Should go to page 2 and selected number of rows should be the same number', () => {
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:1374]:
```suggestion
    it('The number of selected rows should remain the same when navigating to page 2', () => {
```

File: cypress/integration/chi-vue/data-table.spec.js [hunk 12]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+                    .${PAGINATION_CLASSES.RESULTS}`
+              )
+                .find('span')
+                .contains(`${paginationResults} results`);
+              expect(spy).to.be.calledOnce;
+            });
+        });
+    });
+
+    it.skip('Should close bulk actions when "Show selected only" is active and show data table is still in its original state', () => {
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:1322]: The text is not descriptive enough.

File: cypress/integration/chi-vue/data-table.spec.js [hunk 13]
```diff
@@ -1175,6 +1190,238 @@ describe('Data Table', () => {
...
+                  .${BULK_ACTIONS_CLASSES.LABEL}`)
+            .contains(`Actions (${selected} Selected)`);
+        });
+      cy.get(`@pagination`)
+        .find(`.${ICON_BUTTON}`)
+        .first()
+        .click();
+    });
+
+    it('Should go to page 2 and selected number of rows should update', () => {
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:1399]:
```suggestion
    it('The number of selected rows should update when selecting rows on page 2', () => {
```

File: cypress/integration/chi-vue/data-table.spec.js [hunk 14]
```diff
@@ -346,25 +362,24 @@ describe('Data Table', () => {
 
     const headCellsTooltips = [false, false, true, true, true];
 
-    headCellsTooltips.forEach((isVisible, index) => {
-      const assertion = !isVisible ? 'not.exist' : 'exist';
+    headCellsTooltips.forEach((existsAssertion, index) => {
+      const assertion = !existsAssertion ? 'not.exist' : 'exist';
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:366]:
```suggestion
    headCellsTooltips.forEach((isVisible, index) => {
      const assertion = !isVisible ? 'not.exist' : 'exist';
```

Comment (janibolkvadze) [reply cypress/integration/chi-vue/data-table.spec.js:366]: @evansito @josecarlosgonzalezv `is` prefixed names are standard for boolean values like this.
`isVisible true`
is more readable than
`existsAssertion true`

File: cypress/integration/chi-vue/data-table.spec.js [hunk 15]
```diff
@@ -346,25 +362,24 @@ describe('Data Table', () => {
...
-          .as('trigger')
-          .trigger('mouseenter')
-          .then(() => {
-            cy.get('@trigger')
-              .find(`.${TOOLTIP_CLASSES.TOOLTIP}`)
-              .should(assertion);
-          });
-      });
+      it(`Tooltip element ${index} should ${!existsAssertion ? 'not' : ''
+        } exist as the label is ${!existsAssertion ? 'not' : ''} truncated`, () => {
```
Comment (janibolkvadze) [inline cypress/integration/chi-vue/data-table.spec.js:369]:
```suggestion
      it(`Tooltip element ${index} should ${!isVisible ? 'not' : ''
        } exist as the label is ${!isVisible ? 'not' : ''} truncated`, () => {
```

**Review by janibolkvadze — CHANGES_REQUESTED** [review]

**Comment by SCMAUTO** [comment]
The CI pipeline run successfully in https://jenkinsprod.corp.intranet:8443/job/UX-CHI/job/Productive/job/Chi/job/PR-314/3/. :white_check_mark:

**Comment by SCMAUTO** [comment]
You can check this PRs instance in https://5536473391251649750ee7a24603a1482056685b-chi-pr.enterprise.ctl.io (external) and in https://nginx-pr-314-ux-chi.kubeodc-test.corp.intranet/ (internal)

**Comment by SCMAUTO** [comment]
The CI pipeline run successfully in https://jenkinsprod.corp.intranet:8443/job/UX-CHI/job/Productive/job/Chi/job/PR-314/4/. :white_check_mark:

**Comment by SCMAUTO** [comment]
You can check this PRs instance in https://05dc4093f68556a0027afb05129fbbd4a5f4b0b2-chi-pr.enterprise.ctl.io (external) and in https://nginx-pr-314-ux-chi.kubeodc-test.corp.intranet/ (internal)

**Review by janibolkvadze — APPROVED** [review]

**Comment by SCMAUTO** [comment]
The CI pipeline run successfully in https://jenkinsprod.corp.intranet:8443/job/UX-CHI/job/Productive/job/Chi/job/PR-314/5/. :white_check_mark:

**Comment by SCMAUTO** [comment]
You can check this PRs instance in https://9a20d8fe59b00e37db036cf02030f8085f0f6db4-chi-pr.enterprise.ctl.io (external) and in https://nginx-pr-314-ux-chi.kubeodc-test.corp.intranet/ (internal)

**Review by evansito — APPROVED** [review]

**Pull request merged**